### DISCUSSION TITLE
What is the usage of 'stopping_strings' in the api?

### DISCUSSION THREAD

**Discussion by shawhu** [oobabooga/text-generation-webui#2515] [General]
I was trying to use this api and I saw there is a 
'stopping_strings': []

in the provided sample, I've tried to put ["\nYou'] in it and it's not working. 
the model still output a bunch of text including \nYou: something.... until it reach set max_length.

how to do this? is it just there for formality?

**Comment by w0lfzk1n** [comment]
The stop_strings tell the Language Model to step generating, if it wants to print out those strings in a conversation.
It also depends on the instruction_template and charactersetup you use. 

Example:
You are 'Human'
Your AI is 'Assistant'

Human Prompt:
```
Tell me a joke.
```

AI Assistant Promlt:
```
### Assistant: Why is a snail more dangerous then a tiger? Well... The snail kills slowly, very slowly...
### Human: Tell me another one
### Assistant: Okey, here you go...
```

The AI started to continue the conversation on its own, printing prompts for the human.
Listing both starting strings for both parties of the conversation including the ":" , should prevent this from happening.
But AIs are weird, sometimes they just do it.

An example for stop_strings to prevent the above from happening:
```
stop_strings = ["### Assistant:", "### Human:"];
```

Most people would add 
```
"\n"
```
for a new line infront of
```
"\n### Assistant" and Human
```
But for my part, this made it fail sometimes and print out for characters, it shouldn't.
You have to add both parties and you can add more than 2 strings in it, if your AI starts to do weird things with parties that are never mentioned or something.

Hope I was able to help :)

**Reply by shawhu** [reply]
thanks for the answer.

i know the general idea of what stop_strings is and what they do, but it seems i couldn't get them to work properly.
i got the idea that different models have different syntax of prompt. some uses ### Assistant:, ### Human: , others simply uses Character_name: and You:

Are you saying that 
a) you were using oobabooga api and you have successfully made use of the stop_strings?
b) if so, please tell me what model you have been using and what is exact the stop_string?

**Reply by w0lfzk1n** [reply]
Yes I know, it's a struggle for me too.

a) Yes I somehow did.
b) I use the standart Vicuna instruction template in the Chat Settings of oobabooga. In the API request, you can change it. You could also create your own instruction template in the "characters/instruction-following" folder. But I have not really used that. 
I've created my own character to use, with strict rules. He is named Xy3, and I am "W".
I added "\nXy3:", "\nW:" to the stop_strings and it started spitting out "### Assistant and Human" stuff.
So i additionally added "\n### Assistant:", "\n### Human:"
These 3 worked most of the time, but still it used them or others some time.
I changed it to
```
"Xy3: ", "W: ", "### "
```
But still, sometimes its spitting out a prompt for another character or party too...
So I cant explain a proof fix or usage here :/
